### PR TITLE
feat(proxy): unify local network proxy behavior

### LINKED ISSUE #10631: [BUG] HTTP requests to localhost not going through proxy with Chrome
**Context:**
- Playwright Version: 1.16.3
- Operating System: Mac and Linux
- Node.js version: v12.20.0
- Browser: Chromium
- Extra: Proxy configuration to test with OWASP

**Code Snippet**

I'm setting the proxy options in the `use` section of `playwright.config.js` like this:

```javascript
// playwright.config.js
import { PlaywrightTestConfig, devices } from '@playwright/test';

const proxyHost = process.env.LOCAL_PROXY_HOST || 'localhost';
const proxyPort = process.env.LOCAL_PROXY_PORT || '8080';

const config: PlaywrightTestConfig = {
  testDir: 'tests',
  forbidOnly: !!process.env.CI,
  retries: process.env.CI ? 2 : 0,
  globalSetup: require.resolve('./global-setup'),
  use: {
    baseURL: 'http://localhost:4200',
    viewport: { width: 1280, height: 1580 },
    ignoreHTTPSErrors: true,
    trace: 'on-first-retry',
    storageState: 'auth-state.json',
    proxy = {
      server: `http://${proxyHost}:${proxyPort}`
    }
  },
  webServer: {
    command: 'npx ng serve',
    port: 4200,
    timeout: 120 * 1000,
  },
  projects: [
    {
      name: 'chromium',
      use: { ...devices['Desktop Chrome'] },
    }
  ],
};
export default config;
```

**Describe the bug**

When running playwright with a simple Angular application that is served from `localhost:4200`, requests to `localhost` are not going through the configured proxy (they are not logged in the proxy) but all other requests do. This problem only occurs with a Chromium browser, it works as expected with Firefox where I can see all requests logged in the proxy.

Is this a known behavior or are there any additional options we could set to enforce proxy for localhost? Browserstack has similar options (`forceProxy`).

### PR DESCRIPTION
**Pull request by rwoll**
When configuring a proxy, Chromium requires a magic tokens to get some
local network requests to go through the proxy. This has tripped up a
few users, so we make the behavior default to the expected: proxy
everything including the local requests. This matches the other vendors
as well.

NB: This can be disabled via
`PLAYWRIGHT_DISABLE_FORCED_CHROMIUM_PROXIED_LOOPBACK=1`

Supercedes: #8345
Fixes: #10631

### CODE REVIEW

**Comment by rwoll** [comment]
Looks like the FF failures have to do with my test, and not PW impl. Looks like our `proxyServer` is too naïve for FF to startup properly (if used at the launch- but not context-level). The following blows up the `newPage` call in a funny way logging `Error: Parse Error: Invalid method encountered`. The only request I see hit the proxy before that is maybe an OCSP related one to http://ocsp.digicert.com/. Need to dig in later…

```
it.only('debug', async ({ browserType, server, proxyServer }) => {
  server.setRoute('/target.html', async (req, res) => {
    res.end('<html><title>Served by the proxy</title></html>');
  });
  proxyServer.forwardTo(server.PORT);
  const browser = await browserType.launch({
    proxy: { server: `localhost:${proxyServer.PORT}` }
  });
  const page = await browser.newPage();
  await browser.close();
});
```

**Comment by rwoll** [comment]
> Looks like the FF failures have to do with my test, and not PW impl. Looks like our `proxyServer` is too naïve for FF to startup properly (if used at the launch- but not context-level). The following blows up the `newPage` call in a funny way logging `Error: Parse Error: Invalid method encountered`. The only request I see hit the proxy before that is maybe an OCSP related one to http://ocsp.digicert.com/. Need to dig in later…
> 
> ```
> it.only('debug', async ({ browserType, server, proxyServer }) => {
>   server.setRoute('/target.html', async (req, res) => {
>     res.end('<html><title>Served by the proxy</title></html>');
>   });
>   proxyServer.forwardTo(server.PORT);
>   const browser = await browserType.launch({
>     proxy: { server: `localhost:${proxyServer.PORT}` }
>   });
>   const page = await browser.newPage();
>   await browser.close();
> });
> ```

Looks like the issue is with trying to proxy the HTTPS/connect requests in proxyServer which originate from browser startup.

**Comment by rwoll** [comment]
> > Looks like the FF failures have to do with my test, and not PW impl. Looks like our `proxyServer` is too naïve for FF to startup properly (if used at the launch- but not context-level). The following blows up the `newPage` call in a funny way logging `Error: Parse Error: Invalid method encountered`. The only request I see hit the proxy before that is maybe an OCSP related one to http://ocsp.digicert.com/. Need to dig in later…
> > 
> > ```
> > it.only('debug', async ({ browserType, server, proxyServer }) => {
> >   server.setRoute('/target.html', async (req, res) => {
> >     res.end('<html><title>Served by the proxy</title></html>');
> >   });
> >   proxyServer.forwardTo(server.PORT);
> >   const browser = await browserType.launch({
> >     proxy: { server: `localhost:${proxyServer.PORT}` }
> >   });
> >   const page = await browser.newPage();
> >   await browser.close();
> > });
> > ```
> 
> Looks like the issue is with trying to proxy the HTTPS/connect requests in proxyServer which originate from browser startup.

Pushed a workaround in latest commit. (Skips rewriting some HTTPS requests FF requires for startup so we don't have to deal with MITMing them better just for the tests.)

File: tests/browsercontext-proxy.spec.ts
Comment (pavelfeldman) [inline tests/browsercontext-proxy.spec.ts:94]: nit: we are going to bulk-enable parallel for our suite, so no need to opt-in.

File: tests/browsercontext-proxy.spec.ts
Comment (pavelfeldman) [inline tests/browsercontext-proxy.spec.ts:112]: nit, style: given the complexity of the fail predicate, I'd copy-paste this test 3 times instead.

**Review by pavelfeldman — APPROVED** [review]

File: tests/browsercontext-proxy.spec.ts
Comment (rwoll) [reply tests/browsercontext-proxy.spec.ts]: > nit, style: given the complexity of the fail predicate, I'd copy-paste this test 3 times instead.

@pavelfeldman how strong of a nit is this? Hehe 

I found the more complex fail predicate much easier to read/maintain/edit than trying to find the subtle differences in the otherwise lots of setup in each of the cases. 

Mind if I revert https://github.com/microsoft/playwright/commit/d307a872d3a5a432f82fb88cc71c63c50b03e01c before merging?

Comment (pavelfeldman) [reply tests/browsercontext-proxy.spec.ts]: > how strong of a nit is this?

Not too strong, do it the way you like.

Comment (pavelfeldman) [reply tests/browsercontext-proxy.spec.ts]: Happy to merge as is if this is ready.

Comment (rwoll) [reply tests/browsercontext-proxy.spec.ts:112]: Thanks @pavelfeldman! Just reverted! Ready for a merge whenever assuming tests pass.

**Comment by rwoll** [comment]
Fixed up the tests based on the CQ1 run. Ready for another run. (Ran all the proxy tests locally {macOS, Linux/Docker} x {headless, head full}. Don’t have a windows box to verifies this will pass now, but they should :)

**Comment by rwoll** [comment]
@mxschmitt Thanks for toggling the label. Looks like `npm ci` exploded on a lot of the bots and didn’t even run the tests. Is that just runner flakiness, or should I rebase and try again?

**Comment by mxschmitt** [comment]
> @mxschmitt Thanks for toggling the label. Looks like `npm ci` exploded on a lot of the bots and didn’t even run the tests. Is that just runner flakiness, or should I rebase and try again?

Seems like a temporary GitHub/NPM outage, I'd recommend to rebase.

**Comment by rwoll** [comment]
Thanks for the re-trigger! Everything passed on the rebase except one unrelated test (https://github.com/microsoft/playwright/runs/4464958397?check_suite_focus=true)

So should be good to merge! 🎉 

cc: @pavelfeldman

**Pull request merged**
